# Comparing `tmp/genofunc-1.1.2-py3-none-any.whl.zip` & `tmp/genofunc-1.1.3-py3-none-any.whl.zip`

## zipinfo {}

```diff
@@ -1,36 +1,36 @@
-Zip file size: 28965 bytes, number of entries: 34
+Zip file size: 28962 bytes, number of entries: 34
 -rw-r--r--  2.0 unx      537 b- defN 22-Mar-25 16:35 genofunc/__init__.py
 -rw-r--r--  2.0 unx    16649 b- defN 23-May-18 10:01 genofunc/__main__.py
 -rw-r--r--  2.0 unx     1419 b- defN 22-Mar-25 16:35 genofunc/concatenate_fasta.py
 -rw-r--r--  2.0 unx     2932 b- defN 22-Mar-25 16:35 genofunc/extract_metadata.py
 -rw-r--r--  2.0 unx     3781 b- defN 23-May-11 11:44 genofunc/feature_extractor.py
 -rw-r--r--  2.0 unx     4847 b- defN 22-Mar-25 16:35 genofunc/filter_fasta.py
 -rw-r--r--  2.0 unx     1576 b- defN 22-Mar-25 16:35 genofunc/gene_concatenator.py
 -rw-r--r--  2.0 unx     6411 b- defN 23-May-19 11:58 genofunc/genome_annotator.py
 -rw-r--r--  2.0 unx     7691 b- defN 22-Jun-10 10:33 genofunc/genome_annotator2.py
 -rw-r--r--  2.0 unx     4677 b- defN 22-Mar-25 16:35 genofunc/group_align.py
 -rw-r--r--  2.0 unx     4106 b- defN 22-Mar-25 16:35 genofunc/merge.py
 -rw-r--r--  2.0 unx     1587 b- defN 22-Mar-25 16:35 genofunc/name_splitter.py
 -rw-r--r--  2.0 unx     3924 b- defN 23-May-15 10:21 genofunc/reference_matcher.py
 -rw-r--r--  2.0 unx     1521 b- defN 22-Mar-25 16:35 genofunc/rename_fasta.py
--rw-r--r--  2.0 unx     2789 b- defN 22-Mar-25 16:35 genofunc/strain_encoder.py
+-rw-r--r--  2.0 unx     2677 b- defN 23-Jun-03 11:35 genofunc/strain_encoder.py
 -rw-r--r--  2.0 unx      650 b- defN 22-Mar-25 16:35 genofunc/utils.py
 -rw-r--r--  2.0 unx      410 b- defN 22-Mar-25 16:35 genofunc/subcommands/__init__.py
 -rw-r--r--  2.0 unx      350 b- defN 22-Mar-25 16:35 genofunc/subcommands/concatenate_fasta.py
 -rw-r--r--  2.0 unx      536 b- defN 22-Mar-25 16:35 genofunc/subcommands/extract_metadata.py
 -rw-r--r--  2.0 unx      418 b- defN 22-Mar-25 16:35 genofunc/subcommands/feature_extractor.py
 -rw-r--r--  2.0 unx      448 b- defN 22-Mar-25 16:35 genofunc/subcommands/filter_fasta.py
 -rw-r--r--  2.0 unx      308 b- defN 22-Mar-25 16:35 genofunc/subcommands/gene_concatenator.py
 -rw-r--r--  2.0 unx      359 b- defN 23-May-18 10:02 genofunc/subcommands/genome_annotator.py
 -rw-r--r--  2.0 unx      342 b- defN 22-Mar-25 16:35 genofunc/subcommands/group_align.py
 -rw-r--r--  2.0 unx      325 b- defN 22-Mar-25 16:35 genofunc/subcommands/merge.py
 -rw-r--r--  2.0 unx      373 b- defN 22-Mar-25 16:35 genofunc/subcommands/name_splitter.py
 -rw-r--r--  2.0 unx      391 b- defN 23-May-10 09:17 genofunc/subcommands/reference_matcher.py
 -rw-r--r--  2.0 unx      317 b- defN 22-Mar-25 16:35 genofunc/subcommands/rename_fasta.py
 -rw-r--r--  2.0 unx      397 b- defN 22-Mar-25 16:35 genofunc/subcommands/strain_encoder.py
--rw-r--r--  2.0 unx      572 b- defN 23-May-19 12:03 genofunc-1.1.2.dist-info/METADATA
--rw-r--r--  2.0 unx       92 b- defN 23-May-19 12:03 genofunc-1.1.2.dist-info/WHEEL
--rw-r--r--  2.0 unx       53 b- defN 23-May-19 12:03 genofunc-1.1.2.dist-info/entry_points.txt
--rw-r--r--  2.0 unx        9 b- defN 23-May-19 12:03 genofunc-1.1.2.dist-info/top_level.txt
-?rw-rw-r--  2.0 unx     2919 b- defN 23-May-19 12:03 genofunc-1.1.2.dist-info/RECORD
-34 files, 73716 bytes uncompressed, 24245 bytes compressed:  67.1%
+-rw-r--r--  2.0 unx      572 b- defN 23-Jun-03 11:39 genofunc-1.1.3.dist-info/METADATA
+-rw-r--r--  2.0 unx       92 b- defN 23-Jun-03 11:39 genofunc-1.1.3.dist-info/WHEEL
+-rw-r--r--  2.0 unx       53 b- defN 23-Jun-03 11:39 genofunc-1.1.3.dist-info/entry_points.txt
+-rw-r--r--  2.0 unx        9 b- defN 23-Jun-03 11:39 genofunc-1.1.3.dist-info/top_level.txt
+?rw-rw-r--  2.0 unx     2919 b- defN 23-Jun-03 11:39 genofunc-1.1.3.dist-info/RECORD
+34 files, 73604 bytes uncompressed, 24242 bytes compressed:  67.1%
```

## zipnote {}

```diff
@@ -81,23 +81,23 @@
 
 Filename: genofunc/subcommands/rename_fasta.py
 Comment: 
 
 Filename: genofunc/subcommands/strain_encoder.py
 Comment: 
 
-Filename: genofunc-1.1.2.dist-info/METADATA
+Filename: genofunc-1.1.3.dist-info/METADATA
 Comment: 
 
-Filename: genofunc-1.1.2.dist-info/WHEEL
+Filename: genofunc-1.1.3.dist-info/WHEEL
 Comment: 
 
-Filename: genofunc-1.1.2.dist-info/entry_points.txt
+Filename: genofunc-1.1.3.dist-info/entry_points.txt
 Comment: 
 
-Filename: genofunc-1.1.2.dist-info/top_level.txt
+Filename: genofunc-1.1.3.dist-info/top_level.txt
 Comment: 
 
-Filename: genofunc-1.1.2.dist-info/RECORD
+Filename: genofunc-1.1.3.dist-info/RECORD
 Comment: 
 
 Zip file comment:
```

## genofunc/strain_encoder.py

```diff
@@ -1,13 +1,13 @@
 #!/usr/bin/env python3
 
 """
 Name: strain_encoder.py
 Author: Xiaoyu Yu
-Date: 01 April 2022
+Date: 03 June 2023
 Description: Encoded strain id into non-defining ids. 
 
 Options:
     :param in_fasta: Input folder containing alignments in fasta format (Required)
     :param in_metadata: Input metadata for cohort information (Required)
     :param encoding_column: Column for the base for encoding information (Required)
     :param out_dir: Output folder including encoded files (Default: ./)
@@ -30,24 +30,23 @@
         print("No fasta files in input directory. Please re-enter new one. Program Exiting.")
         sys.exit()
 
     if not file_check(in_metadata):
         print("Metadata file does not exist. Please enter a new file path. Program Exiting")
         sys.exit()
 
-
-    output_metadata = open(out_dir+"encoded_metadata.csv","w")
     log_handle = get_log_handle(log_file)
     column_dic = {}
     encoding_dic = {}
 
     with open(in_metadata,"r") as f:
         seperator = ','
         if in_metadata.endswith('tsv'):
             seperator = '\t'
+        header = f.readline().split(seperator)
         for line in f:
             rows = line.rstrip().split(seperator)
             if rows[encoding_column] not in column_dic.keys():
                 column_dic[rows[encoding_column]] = [rows[0]]
             else:
                 column_dic[rows[encoding_column]].append(rows[0])
 
@@ -62,21 +61,17 @@
         output_sequence = open(out_dir+file_name[:-6]+"_encoded.fasta","w")
         for record in SeqIO.parse(files, "fasta"):
             new_id = encoding_dic[record.id]
             new_record = SeqRecord(record.seq,new_id,description="")
             SeqIO.write(new_record, output_sequence, "fasta-2line")
         output_sequence.close() 
 
+    output_metadata = open(out_dir+"encoded_metadata.csv","w")
     with open(in_metadata,"r") as f:
-        seperator = ','
-        if in_metadata.endswith('tsv'):
-            seperator = '\t'
         for line in f:
             rows = line.rstrip().split(seperator)
-            if rows[0] == "strain":
-                output_metadata.write(line)
-                continue
-            rows[0] = encoding_dic[rows[0]]
+            if rows[0] != header[0]:
+                rows[0] = encoding_dic[rows[0]]
             output_metadata.write(",".join(rows)+"\n")
 
     close_handle(output_metadata)
     close_handle(log_handle)
```

## Comparing `genofunc-1.1.2.dist-info/METADATA` & `genofunc-1.1.3.dist-info/METADATA`

 * *Files 18% similar despite different names*

```diff
@@ -1,10 +1,10 @@
 Metadata-Version: 2.1
 Name: genofunc
-Version: 1.1.2
+Version: 1.1.3
 Summary: UNKNOWN
 Home-page: https://github.com/xiaoyu518/genofunc
 Author: UNKNOWN
 Author-email: UNKNOWN
 License: MIT
 Platform: UNKNOWN
 Classifier: Topic :: Scientific/Engineering :: Bio-Informatics
```

## Comparing `genofunc-1.1.2.dist-info/RECORD` & `genofunc-1.1.3.dist-info/RECORD`

 * *Files 16% similar despite different names*

```diff
@@ -8,27 +8,27 @@
 genofunc/genome_annotator.py,sha256=XwSlmXoJUprEkYWcpbh-KNRckUTY-92fN2mwUiMubZE,6411
 genofunc/genome_annotator2.py,sha256=4PUGUaZKJYd-DJIPhHcKYvyLVT7fbKMlqAalktKz4tM,7691
 genofunc/group_align.py,sha256=qj0jnLN3v-sE1UjApBDHosGRlfcdBCoNKBrjfcvEErQ,4677
 genofunc/merge.py,sha256=XHb626y6u5plNmcfejRXmgzO3D17pW7J4nC96PnvYUk,4106
 genofunc/name_splitter.py,sha256=LMvmfMCSGSHKrShR-bIim5nALq4Q1dKV30X4xpCnIH4,1587
 genofunc/reference_matcher.py,sha256=FnhCDGcDiFXK5ZGETVw5y3-Dvs4d4z2ZdkUzYSAq9yQ,3924
 genofunc/rename_fasta.py,sha256=RpIto02NkmdxnpcbPL2rlw69MZ6jMm4Lrx4aq3Q08bA,1521
-genofunc/strain_encoder.py,sha256=tCwzxo5V9Y9_JHApxfL0RX8wg0epgsHk4yBDbX_N_Jc,2789
+genofunc/strain_encoder.py,sha256=uXflOAplmr43jllJiP2AR1bCQB9BgjTResVl6MlkMzs,2677
 genofunc/utils.py,sha256=K0KHBLHCTm9Kgv07IpSllSfHyNzjKBOgPIO7ZYdl4Ng,650
 genofunc/subcommands/__init__.py,sha256=ULmKBq-2JqrAJrh0LKy6yDG_7a4GeTmE7sXjPEmQOHM,410
 genofunc/subcommands/concatenate_fasta.py,sha256=3JD1N14vgcbbAWYKRi725x3OwaI7DLUimbd6_WahGoc,350
 genofunc/subcommands/extract_metadata.py,sha256=ITzNfDsG2WyPhSX1P_SGy6Z57tsJLAjg6O9vIuHp7BM,536
 genofunc/subcommands/feature_extractor.py,sha256=NBTwaasbkeKU2bo3wFqENSK_mwuiO5NYuSIKzVAl8h0,418
 genofunc/subcommands/filter_fasta.py,sha256=4Hdj-34OnhliBn43wK9OBXQkHYU2RJfW6tieWhkMwbE,448
 genofunc/subcommands/gene_concatenator.py,sha256=SIt0jOl-cGE5SukmKLlwPm6hVFUpMzLIMPk9ICPIpno,308
 genofunc/subcommands/genome_annotator.py,sha256=kesTdOZ5l_j2haJdSr4SAbUH1S43k7IozH4h1yQeC6s,359
 genofunc/subcommands/group_align.py,sha256=4dERT7cdM_pgHANrE1WMpBPF03sn11rk0csETmrIU5A,342
 genofunc/subcommands/merge.py,sha256=uTz_oVJ1Tzd958jttdPo8ka3jWIKc0dt4quEuwu2VIw,325
 genofunc/subcommands/name_splitter.py,sha256=s3SMtbRsODSx0RuK6wcByeUJfpPgYLfdMqI1nMJaJYs,373
 genofunc/subcommands/reference_matcher.py,sha256=ArmUvL6JFD8NV2gBMdZrNDnQ4Szby0QLSXKPEQ1oO9w,391
 genofunc/subcommands/rename_fasta.py,sha256=TACml7SbX7ZRRVbJoGNW49I6HdjBxAwj3TVi3N2ClsM,317
 genofunc/subcommands/strain_encoder.py,sha256=wBtrfxFMERiDAEtAziu145M1QChp6prDpGi9RhA5MSk,397
-genofunc-1.1.2.dist-info/METADATA,sha256=Y8uv0_z7gTeZgg8_GTWSNHVsBomRdbLuqMi4wQfunJM,572
-genofunc-1.1.2.dist-info/WHEEL,sha256=U88EhGIw8Sj2_phqajeu_EAi3RAo8-C6zV3REsWbWbs,92
-genofunc-1.1.2.dist-info/entry_points.txt,sha256=9xT6cQgMyyC1ktKvIEtsvI1NFjO_W9iloZ9NbIcPw6Q,53
-genofunc-1.1.2.dist-info/top_level.txt,sha256=rdj71U7IfsnpHsF93VlZG0FTDrPHDxaua4sx_F5yYUI,9
-genofunc-1.1.2.dist-info/RECORD,,
+genofunc-1.1.3.dist-info/METADATA,sha256=AXXD4U9a6N80GnUmM6zLiuIUvCmuRlnCt6pAvXiA4b8,572
+genofunc-1.1.3.dist-info/WHEEL,sha256=U88EhGIw8Sj2_phqajeu_EAi3RAo8-C6zV3REsWbWbs,92
+genofunc-1.1.3.dist-info/entry_points.txt,sha256=9xT6cQgMyyC1ktKvIEtsvI1NFjO_W9iloZ9NbIcPw6Q,53
+genofunc-1.1.3.dist-info/top_level.txt,sha256=rdj71U7IfsnpHsF93VlZG0FTDrPHDxaua4sx_F5yYUI,9
+genofunc-1.1.3.dist-info/RECORD,,
```

